vnext/ReactWindowsCore/WinRTWebSocketResource.cpp

```diff
@@ -0,0 +1,363 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+// clang-format off
+#include "WinRTWebSocketResource.h"
+
+#include <winrt/Windows.Security.Cryptography.h>
+#include <Unicode.h>
+#include <Utilities.h>
+
+// Standard Library
+#include <future>
+
+using Microsoft::Common::Unicode::Utf8ToUtf16;
+using Microsoft::Common::Unicode::Utf16ToUtf8;
+using Microsoft::Common::Utilities::CheckedReinterpretCast;
+
+using std::function;
+using std::size_t;
+using std::string;
+using std::vector;
+
+using winrt::fire_and_forget;
+using winrt::hresult;
+using winrt::hresult_error;
+using winrt::resume_background;
+using winrt::Windows::Foundation::Uri;
+using winrt::Windows::Networking::Sockets::IWebSocket;
+using winrt::Windows::Networking::Sockets::MessageWebSocket;
+using winrt::Windows::Networking::Sockets::MessageWebSocketMessageReceivedEventArgs;
+using winrt::Windows::Networking::Sockets::SocketMessageType;
+using winrt::Windows::Networking::Sockets::WebSocketClosedEventArgs;
+using winrt::Windows::Security::Cryptography::CryptographicBuffer;
+using winrt::Windows::Storage::Streams::DataReader;
+using winrt::Windows::Storage::Streams::DataWriter;
+using winrt::Windows::Storage::Streams::UnicodeEncoding;
+
+namespace Microsoft::React
+{
+WinRTWebSocketResource::WinRTWebSocketResource(Uri&& uri)
+  : m_uri{ std::move(uri) }
+  , m_socket{}
+  , m_writer{ m_socket.OutputStream() }
+  , m_closeCode{ CloseCode::None }
+{
+  m_socket.MessageReceived({ this, &WinRTWebSocketResource::OnMessageReceived });
+  m_socket.Closed({ this, &WinRTWebSocketResource::OnClosed });
+
+  m_socket.Control().IgnorableServerCertificateErrors()
+    .Append(winrt::Windows::Security::Cryptography::Certificates::ChainValidationResult::Untrusted);
+  m_socket.Control().IgnorableServerCertificateErrors()
+    .Append(winrt::Windows::Security::Cryptography::Certificates::ChainValidationResult::InvalidName);
+
+}
+
+WinRTWebSocketResource::WinRTWebSocketResource(const string& urlString)
+  : WinRTWebSocketResource(Uri{ Utf8ToUtf16(urlString) })
+{
+}
+
+WinRTWebSocketResource::~WinRTWebSocketResource() /*override*/
+{
+  Stop();
```

Comment:
`Stop()` currently blocks the thread. Having this as the side-effect of a destructor causes potentially unexpected behavior, where something holding onto this can cause hangs without realizing it. We should generally try to make long operations really obvious. This conflicts a bit with RAII, but a strategy to force callers to deal with this type of thing is to 

1. Have an explicit method that does the long thing they need to plan around
2. Assert in the destructor the thing happened instead of just doing it

Reply:
There are scenarios where the objects are destroyed without having an explicit call to `Close`.
This class must be able to clean up after itself when being destroyed.

The mechanism is similar to `BeastWebSocketResource`.